marshal/cluster.go

```diff
@@ -0,0 +1,272 @@
+	}
+
+	// Now we start a goroutine to start consuming each of the partitions in the marshal
+	// topic. Note that this doesn't handle increasing the partition count on that topic
+	// without stopping all consumers.
+	c.rationalizers.Add(c.partitions)
+	for id := 0; id < c.partitions; id++ {
+		go c.rationalize(id, c.kafkaConsumerChannel(id))
+	}
+
+	// A jitter calculator, just fills a channel with random numbers so that other
+	// people don't have to build their own random generator...
+	go func() {
+		rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+		for {
+			jitter := rnd.Intn(HeartbeatInterval/2) + (HeartbeatInterval / 2)
+			c.jitters <- time.Duration(jitter) * time.Second
+		}
+	}()
+
+	// Now start the metadata refreshing goroutine
+	go func() {
+		for !c.terminated() {
+			time.Sleep(<-c.jitters)
+			log.Infof("Refreshing topic metadata.")
+			c.refreshMetadata()
+
+			// See if the number of partitions in the marshal topic changed. This is bad if
+			// it happens, since it means we can no longer coordinate correctly.
+			if c.getTopicPartitions(MarshalTopic) != c.partitions {
+				log.Errorf("Marshal topic partition count changed. Terminating!")
+				c.terminate()
+			}
+		}
+	}()
+
+	// Wait for all rationalizers to come alive
+	log.Infof("Waiting for all rationalizers to come alive.")
+	c.rationalizers.Wait()
+	log.Infof("All rationalizers alive, Marshaler now alive.")
+
+	return c, nil
+}
+
+// refreshMetadata is periodically used to update our internal state with topic information
+// about the world.
+func (c *cluster) refreshMetadata() error {
+	md, err := c.broker.Metadata()
+	if err != nil {
+		return err
+	}
+
+	newTopics := make(map[string]int)
+	for _, topic := range md.Topics {
+		newTopics[topic.Name] = len(topic.Partitions)
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.topics = newTopics
+	return nil
+}
+
+// getOffsetCoordinator returns a kafka.OffsetCoordinator for a specific group.
+func (c *cluster) getOffsetCoordinator(groupID string) (kafka.OffsetCoordinator, error) {
+	return c.broker.OffsetCoordinator(
+		kafka.NewOffsetCoordinatorConf(groupID))
+}
+
+// getClaimPartition calculates which partition a topic should use for coordination. This uses
+// a hashing function (non-cryptographic) to predictably partition the topic space.
+func (c *cluster) getClaimPartition(topicName string) int {
+	// We use MD5 because it's a fast and good hashing algorithm and we don't need cryptographic
+	// properties. We then take the first 8 bytes and treat them as a uint64 and modulo that
+	// across how many partitions we have.
+	hash := md5.Sum([]byte(topicName))
+	uval := binary.LittleEndian.Uint64(hash[0:8])
+	return int(uval % uint64(c.partitions))
+}
+
+// getPartitionState returns a topicState and possibly creates it and the partition state within
+// the State.
+func (c *cluster) getPartitionState(groupID, topicName string, partID int) *topicState {
```

Comment:
Can you confirm this function and the ones nearby didn't change (just a copy/paste) ? So I can be lazy and not review them lol